sdk/python/lib/pulumi/dynamic/__main__.py

```diff
@@ -36,13 +39,23 @@
 _GRPC_CHANNEL_OPTIONS = [("grpc.max_receive_message_length", _MAX_RPC_MESSAGE_SIZE)]
 
 
-_PROVIDER_CACHE: typing.Dict[str, ResourceProvider] = {}
+_PROVIDER_CACHE: Dict[str, ResourceProvider] = {}
 _PROVIDER_LOCK = Lock()
 
 
-def get_provider(props) -> ResourceProvider:
+def get_provider(props: Dict[str, Any], config: Dict[str, Any]) -> ResourceProvider:
```

Comment:
Maybe a small doc comment on this to call out the lifecycle that it's encapsulating?